src/supa/documents/topology.py

```diff
@@ -0,0 +1,153 @@
+from supa.util.timestamp import current_timestamp
+
+log = structlog.get_logger(__name__)
+
+
+def refresh_topology() -> None:
+    """Refresh list of STP's in the database with topology from NRM, skip if configured for manual topology."""
+    if settings.manual_topology:
+        log.debug("skipping topology refresh", manual_topology=settings.manual_topology)
+        return
+
+    from supa.db.session import db_session
+
+    nrm_stps = get_topology()
+    nrm_stp_ids = [nrm_stp.stp_id for nrm_stp in nrm_stps]
+    with db_session() as session:
+        for nrm_stp in nrm_stps:
+            if nrm_stp.topology != settings.topology:
+                log.debug("skip STP with unknown topology", stp=nrm_stp.stp_id, topology=nrm_stp.topology)
+            else:
+                stp = session.query(Topology).filter(Topology.stp_id == nrm_stp.stp_id).one_or_none()
+                if stp:
+                    log.debug("update existing STP", stp=nrm_stp.stp_id)
+                    stp.port_id = nrm_stp.port_id
+                    stp.vlans = nrm_stp.vlans
+                    stp.description = nrm_stp.description
+                    stp.is_alias_in = nrm_stp.is_alias_in
+                    stp.is_alias_out = nrm_stp.is_alias_out
+                    stp.bandwidth = nrm_stp.bandwidth
+                    stp.enabled = nrm_stp.expose_in_topology
+                else:
+                    log.info("add new STP", stp=nrm_stp.stp_id)
+                    session.add(
+                        Topology(
+                            stp_id=nrm_stp.stp_id,
+                            port_id=nrm_stp.port_id,
+                            vlans=nrm_stp.vlans,
+                            description=nrm_stp.description,
+                            is_alias_in=nrm_stp.is_alias_in,
+                            is_alias_out=nrm_stp.is_alias_out,
+                            bandwidth=nrm_stp.bandwidth,
+                            enabled=nrm_stp.expose_in_topology,
+                        )
+                    )
+        for stp in session.query(Topology).filter(Topology.enabled):
+            if stp.stp_id not in nrm_stp_ids:
+                log.info("disable vanished STP", stp=stp.stp_id)
+                stp.enabled = False
+
+
+"""Namespace map for topology document."""
+nsmap = {
+    "topo": "http://schemas.ogf.org/nml/2013/05/base#",
+    "sd": "http://schemas.ogf.org/nsi/2013/12/services/definition",
+}
+
+
+class TopologyEndpoint(object):
+    """A cherryPy application to generate a NSI topology document."""
+
+    @cherrypy.expose  # type: ignore[misc]
+    def topology(self) -> Union[str, bytes]:
+        """Cherrypy URL that returns the generated NSI topology document."""
+        refresh_topology()
+        network_id = f"urn:ogf:network:{settings.domain}:{settings.topology}"
+        now = current_timestamp()
+
+        from supa.db.session import db_session
+
+        with db_session() as session:
```

Comment:
Does this create a new DB session every time the endpoint is called?

Reply:
Since it's sqlite, it's just opening a file which is inexpensive. But if (as suggested by Rene) supa moves to Postgres then this would be something to improve i.e. by making it setup a connection on application startup

Reply:
The current auto commit is very developer friendly, but the additional overhead of creating a new session every time is something to take into account.